dask/array/core.py

```diff
@@ -894,17 +896,17 @@ def store(sources, targets, lock=True, regions=None, compute=True, **kwargs):
         update = insert_to_ooc(src, tgt, lock=lock, region=reg)
         keys.extend(update)
 
+        update.update(src.dask)
         update.update(dsk)
         updates.update(update)
 
     name = 'store-' + tokenize(*keys)
-    dsk = sharedict.merge((name, updates), *[src.dask for src in sources])
+    dsk = sharedict.merge({name: keys}, updates)
+    result = Delayed(name, dsk)
     if compute:
-        compute_as_if_collection(Array, dsk, keys, **kwargs)
+        result.compute()
     else:
-        from ..delayed import Delayed
-        dsk.update({name: keys})
-        return Delayed(name, dsk)
```

Comment:
This code down here was likely also flawed previously.  We might call the optimize machinery as we do in `Array.to_delayed`

```python
    def to_delayed(self):
        """ Convert Array into dask Delayed objects

        Returns an array of values, one value per chunk.

        See Also
        --------
        dask.array.from_delayed
        """
        from ..delayed import Delayed
        keys = self.__dask_keys__()
        dsk = self.__dask_optimize__(self.__dask_graph__(), keys)
        L = ndeepmap(self.ndim, lambda k: Delayed(k, dsk), keys)
        return np.array(L, dtype=object)
```